src/Eshopworld.WorkerProcess/LeaseAllocator.cs

```diff
@@ -34,83 +35,103 @@ public class LeaseAllocator : ILeaseAllocator
+                _options.Value.Priority, $"Lease allocation started for instance {instanceId}"));
+           
             var lease = await _leaseStore.ReadByLeaseTypeAsync(_options.Value.WorkerType).ConfigureAwait(false);
+            if (lease?.LeasedUntil != null && lease.LeasedUntil.Value > ServerDateTime.UtcNow)
+                return null;
+            
+            await _leaseStore.AddLeaseRequestAsync(_options.Value.WorkerType,_options.Value.Priority,instanceId).ConfigureAwait(false);
 
-            if (lease == null)
-            {
-                var leaseResult = await _leaseStore.TryCreateLeaseAsync(_options.Value.WorkerType,
-                    _options.Value.Priority, instanceId).ConfigureAwait(false);
+            // backoff to allow other workers to add their lease request
+            var delay = _allocationDelay.Calculate(_options.Value.Priority, _options.Value.LeaseInterval);
+            await Task.Delay(delay);
 
-                if (leaseResult.Result)
-                {
-                    lease = leaseResult.Lease;
-                }
-                else
-                {
-                    // another worker lease created the lease before this worker read the created lease
-                    lease = await _leaseStore.ReadByLeaseTypeAsync(_options.Value.WorkerType).ConfigureAwait(false);
-                }
-            }
+            _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
+                _options.Value.Priority, $"Leader election starting"));
 
-            // check lease priority is less than current worker priority and lease is not acquired
-            if (lease.Priority <= _options.Value.Priority && lease.LeasedUntil == null ||
-                lease.LeasedUntil.HasValue && lease.LeasedUntil.Value < ServerDateTime.UtcNow)
-            {
-                lease.Priority = _options.Value.Priority;
-                lease.InstanceId = instanceId;
+            var winnerInstanceId = await _leaseStore.SelectWinnerRequestAsync(_options.Value.WorkerType);
 
-                var updateResult = await _leaseStore.TryUpdateLeaseAsync(lease).ConfigureAwait(false);
+            if (!winnerInstanceId.HasValue || winnerInstanceId != instanceId)
+            {
+                var reason = !winnerInstanceId.HasValue
+                    ? "There are no LeaseRequests to evaluate"
+                    : $"Lease failed to be acquired after election. Winner instance is {winnerInstanceId.Value}.";
+                _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
+                    _options.Value.Priority, reason));
+                return null;
+            }
 
-                if (updateResult.Result)
+            if (lease == null)
+            {
+                var createTime = ServerDateTime.UtcNow;
+                var interval = _slottedInterval.Calculate(createTime, _options.Value.LeaseInterval);
+                lease = new CosmosDbLease
```

Comment:
I think there should be a method that allows you to create a new lease based on a series of parameters (or you should have a class that belongs to this layer - is `CosmosDbLease` not owned by DB layer?

Reply:
Created a new Lease class in a new Model folder. Question: shouldn't ILease as well part of the model as well? currently model has dependency on datalayer (through this interface). It should be the other way around in my opinion at least.